### PR TITLE
feat: [story/EP-1790] new org API

### PR DESCRIPTION
**Pull request by dan-qc**

### CODE REVIEW

File: src/routers/OrganizationRouter.ts [hunk 1]
```diff
@@ -122,6 +124,212 @@ const getAdminRouter = (basePath: string = '/', routePath: string = '/management
...
+      const editorGroupName = [groupName, 'EDITOR'].join('-');
+      const adminGroupName = [groupName, 'ADMIN'].join('-');
+      const ownerGroupName = [groupName, 'OWNER'].join('-');
+
+      // create the main group + nested groups;
+      const root = await authzService.createGroup(groupName, groupDescription);
+      const owner = await authzService.createGroup(ownerGroupName, `${groupName} Owner`);
+      const admin = await authzService.createGroup(adminGroupName, `${groupName} Admin`);
+      const editor = await authzService.createGroup(editorGroupName, `${groupName} Editor`);
+      const viewer = await authzService.createGroup(viewerGroupName, `${groupName} Viewer`);
```
Comment (iacoshoria) [inline src/routers/OrganizationRouter.ts:247]: We can optimize the calls to get them running in parallel:

- using `Promise.all` the resolve order is guaranteed when promises fulfil: 
```
const [root, owner, ...] = Promise.all([
    authzService.createGroup(groupName, groupDescription),
    authzService.createGroup(ownerGroupName, `${groupName} Owner`),
    ...
])

File: src/routers/OrganizationRouter.ts [hunk 2]
```diff
@@ -122,6 +124,212 @@ const getAdminRouter = (basePath: string = '/', routePath: string = '/management
...
+        permissionMap[scope] = permission._id;
+      });
+      await forEachAsync(SCOPES_WRITE, async (scope) => {
+        const write = [groupName, scope].join(':');
+        const permission = await authzService.createPermission(
+          write,
+          SCOPES_WRITE_DESCRIPTION,
+          AUTH0_APPLICATION_CLIENT_ID
+        );
+        permissionMap[scope] = permission._id;
```
Comment (iacoshoria) [inline src/routers/OrganizationRouter.ts:271]: Same as above, we can create all permissions with `Promise.all`

File: src/routers/OrganizationRouter.ts [hunk 3]
```diff
@@ -122,6 +124,212 @@ const getAdminRouter = (basePath: string = '/', routePath: string = '/management
...
+      const ownerName = [groupName, PERMISSIONS.owner.name].join(':');
+      const ownerRole = await authzService.createRole(
+        ownerName,
+        PERMISSIONS.owner.description,
+        AUTH0_APPLICATION_CLIENT_ID,
+        [
+          ...PERMISSIONS.owner.readScopes.map((scope) => permissionMap[scope]),
+          ...PERMISSIONS.owner.writeScopes.map((scope) => permissionMap[scope]),
+        ]
+      );
```
Comment (iacoshoria) [inline src/routers/OrganizationRouter.ts:314]: Same as above, we can create all roles with `Promise.all`

File: src/routers/OrganizationRouter.ts [hunk 4]
```diff
@@ -122,6 +124,212 @@ const getAdminRouter = (basePath: string = '/', routePath: string = '/management
...
+
+      // add the roles for each of the nested groups;
+      await authzService.addGroupRoles(viewer._id, [viewerRole._id]);
+      await authzService.addGroupRoles(editor._id, [editorRole._id]);
+      await authzService.addGroupRoles(admin._id, [adminRole._id]);
+      await authzService.addGroupRoles(owner._id, [ownerRole._id]);
```
Comment (iacoshoria) [inline src/routers/OrganizationRouter.ts:320]: Same as above, we can add the roles to the nested groups with `Promise.all`

File: src/routers/OrganizationRouter.ts [hunk 5]
```diff
@@ -122,6 +124,212 @@ const getAdminRouter = (basePath: string = '/', routePath: string = '/management
...
+
+      const ownerEmail = owners[0];
+
+      const user = await authMgmtService.getUserByEmail(ownerEmail);
+
+      if (!user) {
+        throw new UserNotFoundError('Invalid owner specified.', 404);
+      }
+
+      const SCOPES_READ = [
```
Comment (iacoshoria) [inline src/routers/OrganizationRouter.ts:160]: Might pull this in a separate service, where we can define interfaces for create/delete organizations.

Doesn't need to be handled in this PR.

Comment (dan-qc) [reply src/routers/OrganizationRouter.ts:159]: ok.

**Pull request merged**